File: src/public/StoreTransaction.ts
export class StoreTransaction {
  configRequestId: string;
  appSessionId: string;
  transactionDate?: Date | null;
  originalTransactionIdentifier: string;
  storeTransactionId?: string | null;
  originalTransactionDate?: Date | null;
  webOrderLineItemID?: string | null;
  appBundleId?: string | null;
  subscriptionGroupId?: string | null;
  isUpgraded?: boolean | null;
  expirationDate?: Date | null;
  offerId?: string | null;
  revocationDate?: Date | null;

  constructor({
    configRequestId,
    appSessionId,
    transactionDate,
    originalTransactionIdentifier,
    storeTransactionId,
    originalTransactionDate,
    webOrderLineItemID,
    appBundleId,
    subscriptionGroupId,
    isUpgraded,
    expirationDate,
    offerId,
    revocationDate,
  }: {
    configRequestId: string;
    appSessionId: string;
    transactionDate?: string | null;
    originalTransactionIdentifier: string;
    storeTransactionId?: string | null;
    originalTransactionDate?: string | null;
    webOrderLineItemID?: string | null;
    appBundleId?: string | null;
    subscriptionGroupId?: string | null;
    isUpgraded?: boolean | null;
    expirationDate?: string | null;
    offerId?: string | null;
    revocationDate?: string | null;
  }) {
    this.configRequestId = configRequestId;
    this.appSessionId = appSessionId;
    this.transactionDate = transactionDate ? new Date(transactionDate) : null;
    this.originalTransactionIdentifier = originalTransactionIdentifier;
    this.storeTransactionId = storeTransactionId || null;
    this.originalTransactionDate = originalTransactionDate
      ? new Date(originalTransactionDate)
      : null;
    this.webOrderLineItemID = webOrderLineItemID || null;
    this.appBundleId = appBundleId || null;
    this.subscriptionGroupId = subscriptionGroupId || null;
    this.isUpgraded = isUpgraded ?? null;
    this.expirationDate = expirationDate ? new Date(expirationDate) : null;
    this.offerId = offerId || null;
    this.revocationDate = revocationDate ? new Date(revocationDate) : null;
  }

  static fromJson(json: any): StoreTransaction {
    return new StoreTransaction({
      configRequestId: json.configRequestId,
      appSessionId: json.appSessionId,
      transactionDate: json.transactionDate,
      originalTransactionIdentifier: json.originalTransactionIdentifier,
      storeTransactionId: json.storeTransactionId,
      originalTransactionDate: json.originalTransactionDate,
      webOrderLineItemID: json.webOrderLineItemID,
      appBundleId: json.appBundleId,
      subscriptionGroupId: json.subscriptionGroupId,
      isUpgraded: json.isUpgraded,
      expirationDate: json.expirationDate,
      offerId: json.offerId,
      revocationDate: json.revocationDate,
    });
  }
}

File: android/src/main/java/com/superwallreactnative/models/StoreTransaction.kt
package com.superwallreactnative.models

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.superwall.sdk.store.abstractions.transactions.StoreTransaction
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Date

fun StoreTransaction.toJson(): ReadableMap {
  val map = Arguments.createMap()
  val dateFormatter = DateTimeFormatter.ISO_INSTANT

  fun Date.toIsoString(): String? = this.toInstant()
    .atZone(ZoneId.systemDefault())
    .format(dateFormatter)

  map.putString("configRequestId", configRequestId)
  map.putString("appSessionId", appSessionId)
  map.putString("transactionDate", transactionDate?.toIsoString())
  map.putString("originalTransactionIdentifier", originalTransactionIdentifier)
  map.putString("storeTransactionId", storeTransactionId)
  map.putString("originalTransactionDate", originalTransactionDate?.toIsoString())
  map.putString("webOrderLineItemID", webOrderLineItemID)
  map.putString("appBundleId", appBundleId)
  map.putString("subscriptionGroupId", subscriptionGroupId)
  map.putBoolean("isUpgraded", isUpgraded ?: false)
  map.putString("expirationDate", expirationDate?.toIsoString())
  map.putString("offerId", offerId)
  map.putString("revocationDate", revocationDate?.toIsoString())

  return map
}
